media-proxy/src/libfabric_cq.c

```diff
@@ -78,41 +81,28 @@ void rdma_cq_set_wait_attr(struct fi_cq_attr *cq_attr, enum cq_comp_method metho
     }
 }
 
-static int rdma_spin_for_comp(ep_ctx_t *ep_ctx, struct fid_cq *cq, uint64_t *cur, uint64_t total,
-                              int timeout, struct fi_cq_err_entry *entries)
+static int rdma_spin_for_comp(struct ep_ctx_t *ep_ctx, struct fi_cq_err_entry *entry, int timeout)
 {
-    uint64_t entries_num;
-    struct fi_cq_err_entry comp;
     struct timespec a, b;
     int ret;
 
-    if (!cur)
-        return -EINVAL;
-
-    if (total < *cur)
-        return -EINVAL;
-
-    entries_num = total - *cur;
-
     if (timeout >= 0)
         clock_gettime(CLOCK_MONOTONIC, &a);
 
     do {
-        ret = fi_cq_read(cq, entries ? &entries[entries_num - (total - *cur)] : &comp, 1);
-        if (ret > 0) {
-            if (timeout >= 0)
-                clock_gettime(CLOCK_MONOTONIC, &a);
-            (*cur)++;
-        } else if (ret < 0 && ret != -FI_EAGAIN) {
+        ret = fi_cq_read(ep_ctx->cq_ctx.cq, entry, 1);
+        if (ret < 0 && ret != -FI_EAGAIN) {
             return ret;
-        } else if (timeout >= 0) {
+        } else if (ret == -FI_EAGAIN && timeout >= 0) {
```

Comment:
Not sure about checking for `-FE_EAGAIN` twice here. Maybe slightly rework the flow to have just one check?

Reply:
I rewrote the function. Now it look better